File: packages/python-sdk/src/blindfold/regex/synthesizers.py
"""Format-preserving PII synthesis: generates realistic fake replacements locally."""

import random
import string


def _rand_int(min_val: int, max_val: int) -> int:
    return random.randint(min_val, max_val)


def _rand_char(chars: str) -> str:
    return random.choice(chars)


def _rand_hex(length: int) -> str:
    return "".join(random.choice("0123456789abcdef") for _ in range(length))


def _format_preserving(original: str) -> str:
    """Replace digits with random digits, uppercase with random uppercase,
    lowercase with random lowercase, keep everything else intact."""
    chars = []
    for ch in original:
        if ch.isdigit():
            chars.append(_rand_char(string.digits))
        elif ch.isupper():
            chars.append(_rand_char(string.ascii_uppercase))
        elif ch.islower():
            chars.append(_rand_char(string.ascii_lowercase))
        else:
            chars.append(ch)
    result = "".join(chars)
    # Retry once if we accidentally generated the same value
    if result == original:
        return _format_preserving(original)
    return result


def _synthesize_email() -> str:
    return f"user{_rand_hex(8)}@example.com"


def _synthesize_url() -> str:
    return f"https://example.com/{_rand_hex(10)}"


def _synthesize_ip() -> str:
    a = _rand_int(1, 254)
    b = _rand_int(0, 255)
    c = _rand_int(0, 255)
    d = _rand_int(1, 254)
    return f"{a}.{b}.{c}.{d}"


def _synthesize_credit_card(original: str) -> str:
    """Generate a credit card number with valid Luhn checksum,
    preserving the separator pattern of the original."""
    digits_only = [ch for ch in original if ch.isdigit()]
    num_digits = len(digits_only)

    # Generate random digits (all but the last, which is the Luhn check digit)
    payload = [_rand_int(0, 9) for _ in range(num_digits - 1)]

    # Calculate Luhn check digit
    total = 0
    for i in range(len(payload) - 1, -1, -1):
        d = payload[i]
        if (len(payload) - 1 - i) % 2 == 0:
            d *= 2
            if d > 9:
                d -= 9
        total += d
    check_digit = (10 - (total % 10)) % 10
    payload.append(check_digit)

    # Re-insert separators at the same positions as the original
    digit_idx = 0
    result = []
    for ch in original:
        if ch.isdigit():
            result.append(str(payload[digit_idx]))
            digit_idx += 1
        else:
            result.append(ch)
    return "".join(result)


def _synthesize_iban(original: str) -> str:
    """Generate an IBAN-like value preserving length with a valid mod-97 check."""
    stripped = original.replace(" ", "")
    length = len(stripped)
    if length < 5:
        return _format_preserving(original)

    # Generate random BBAN digits
    bban_len = length - 4
    bban = "".join(_rand_char(string.digits) for _ in range(bban_len))

    # Calculate check digits: convert "XX00" + bban to numeric for mod-97
    # Country letters: X=33, so "XX" -> "3333"
    numeric_str = bban + "333300"
    remainder = 0
    for ch in numeric_str:
        remainder = (remainder * 10 + int(ch)) % 97
    check_digits = str(98 - remainder).zfill(2)

    iban = "XX" + check_digits + bban

    # Re-insert spaces if original had them
    if " " in original:
        space_idx = 0
        result = []
        for ch in original:
            if ch == " ":
                result.append(" ")
            else:
                result.append(iban[space_idx] if space_idx < len(iban) else "0")
                space_idx += 1
        return "".join(result)

    return iban


def synthesize_value(entity_type: str, original_text: str) -> str:
    """Generate a synthetic replacement for a detected PII value.

    Uses format-preserving random by default, with specific generators
    for types where structural validity matters.
    """
    if entity_type == "Email Address":
        return _synthesize_email()
    elif entity_type == "URL":
        return _synthesize_url()
    elif entity_type == "IP Address":
        return _synthesize_ip()
    elif entity_type == "Credit Card Number":
        return _synthesize_credit_card(original_text)
    elif entity_type == "IBAN":
        return _synthesize_iban(original_text)
    else:
        return _format_preserving(original_text)

File: packages/js-sdk/src/regex/synthesizers.ts
// Format-preserving PII synthesis: generates realistic fake replacements locally

const UPPER = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
const LOWER = 'abcdefghijklmnopqrstuvwxyz'
const DIGITS = '0123456789'

function randInt(min: number, max: number): number {
  return Math.floor(Math.random() * (max - min + 1)) + min
}

function randChar(chars: string): string {
  return chars[Math.floor(Math.random() * chars.length)]
}

function randHex(len: number): string {
  let result = ''
  for (let i = 0; i < len; i++) {
    result += randChar('0123456789abcdef')
  }
  return result
}

/**
 * Format-preserving default: replace digits with random digits,
 * uppercase with random uppercase, lowercase with random lowercase,
 * keep everything else (separators, symbols) intact.
 */
function formatPreserving(original: string): string {
  const chars: string[] = []
  for (const ch of original) {
    if (ch >= '0' && ch <= '9') {
      chars.push(randChar(DIGITS))
    } else if (ch >= 'A' && ch <= 'Z') {
      chars.push(randChar(UPPER))
    } else if (ch >= 'a' && ch <= 'z') {
      chars.push(randChar(LOWER))
    } else {
      chars.push(ch)
    }
  }
  const result = chars.join('')
  // Retry once if we accidentally generated the same value
  if (result === original) {
    return formatPreserving(original)
  }
  return result
}

function synthesizeEmail(): string {
  const id = randHex(8)
  return `user${id}@example.com`
}

function synthesizeUrl(): string {
  const path = randHex(10)
  return `https://example.com/${path}`
}

function synthesizeIp(): string {
  const a = randInt(1, 254)
  const b = randInt(0, 255)
  const c = randInt(0, 255)
  const d = randInt(1, 254)
  return `${a}.${b}.${c}.${d}`
}

/**
 * Generate a credit card number with valid Luhn checksum,
 * preserving the separator pattern of the original.
 */
function synthesizeCreditCard(original: string): string {
  // Extract only digits from original to know the length
  const digitsOnly = original.replace(/\D/g, '')
  const numDigits = digitsOnly.length

  // Generate random digits (all but the last, which is the Luhn check digit)
  const payload: number[] = []
  for (let i = 0; i < numDigits - 1; i++) {
    payload.push(randInt(0, 9))
  }

  // Calculate Luhn check digit
  let sum = 0
  for (let i = payload.length - 1; i >= 0; i--) {
    let d = payload[i]
    // Double every second digit from right (the check digit position is even, payload last is odd)
    if ((payload.length - 1 - i) % 2 === 0) {
      d *= 2
      if (d > 9) d -= 9
    }
    sum += d
  }
  const checkDigit = (10 - (sum % 10)) % 10
  payload.push(checkDigit)

  // Re-insert separators at the same positions as the original
  let digitIdx = 0
  let result = ''
  for (const ch of original) {
    if (ch >= '0' && ch <= '9') {
      result += payload[digitIdx++]
    } else {
      result += ch
    }
  }
  return result
}

/**
 * Generate an IBAN-like value preserving length with a valid mod-97 check.
 */
function synthesizeIban(original: string): string {
  // Preserve the length; use XX country code + random digits
  const stripped = original.replace(/\s/g, '')
  const len = stripped.length
  if (len < 5) return formatPreserving(original)

  // Generate random BBAN digits
  const bbanLen = len - 4
  let bban = ''
  for (let i = 0; i < bbanLen; i++) {
    bban += randChar(DIGITS)
  }

  // Calculate check digits: convert "XX00" + bban to numeric for mod-97
  // Country letters: X=33, so "XX" -> "3333"
  const numericStr = bban + '333300'
  // Compute mod-97 on the numeric string in chunks
  let remainder = 0
  for (const ch of numericStr) {
    remainder = (remainder * 10 + parseInt(ch, 10)) % 97
  }
  const checkDigits = (98 - remainder).toString().padStart(2, '0')

  const iban = 'XX' + checkDigits + bban

  // Re-insert spaces if original had them
  if (original.includes(' ')) {
    let spaceIdx = 0
    let result = ''
    for (const ch of original) {
      if (ch === ' ') {
        result += ' '
      } else {
        result += iban[spaceIdx++] || '0'
      }
    }
    return result
  }

  return iban
}

/**
 * Generate a synthetic replacement for a detected PII value.
 * Uses format-preserving random by default, with specific generators
 * for types where structural validity matters.
 */
export function synthesizeValue(entityType: string, originalText: string): string {
  switch (entityType) {
    case 'Email Address':
      return synthesizeEmail()
    case 'URL':
      return synthesizeUrl()
    case 'IP Address':
      return synthesizeIp()
    case 'Credit Card Number':
      return synthesizeCreditCard(originalText)
    case 'IBAN':
      return synthesizeIban(originalText)
    default:
      return formatPreserving(originalText)
  }
}
